### PR TITLE
In-memory cli configuration options; Split `make_ursulas`

### LINKED ISSUE #347: Cleanup on aisle make_ursulas
`make_ursulas` has gotten very messy.  `bare` ? `miners`?  With the event loop at the beginning?  What a cluster.

### PR DESCRIPTION
**Pull request by KPrasch**
Fixes #347

### CODE REVIEW

File: nucypher/config/constants.py [hunk 1]
```diff
@@ -1,3 +1,23 @@
...
+
+import nucypher
+
+# Base Filepaths
+BASE_DIR = abspath(dirname(dirname(nucypher.__file__)))
+PROJECT_ROOT = abspath(dirname(nucypher.__file__))
+APP_DIR = AppDirs("nucypher", "NuCypher")
+
+# Configuration File
+TEMPLATE_CONFIG_FILE_LOCATION = os.path.join(BASE_DIR, 'cli', '.nucypher.ini')
```
Comment (michwill) [inline nucypher/config/constants.py]: We should keep in mind that the `.ini` file will unlikely be alongside the code if `nucypher` is installed with `setup.py` into system directories. Ok for now, but we'll come across this once we start testing with properly installed `nucypher` package

Comment (michwill) [reply nucypher/config/constants.py]: Perhaps, easy enough to fix even in this PR: just need to include `.nucypher.ini` in `package_data` in `setup.py`. We already have Solidity files there

File: nucypher/config/node.py [hunk 2]
```diff
@@ -1,106 +1,100 @@
...
-        self.known_metadata_dir = known_metedata_dir
+
+        if temp is True:
+            self.temp = True
+            self.temp_dir = TemporaryDirectory()
+            self.config_root = self.temp_dir.name
+        else:
+            self.config_root = config_root
+
+        self.operating_mode = 'federated'
```
Comment (michwill) [inline nucypher/config/node.py]: You probably mean `self.operating_mode = operating_mode`?

File: nucypher/utilities/sandbox/blockchain.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    _test_account_cache = list()
+
+    def __init__(self, test_accounts=None, poa=True, airdrop=True, *args, **kwargs):
+
+        # Depends on circumflex
+        super().__init__(*args, **kwargs)
+
+        # For use with Proof-Of-Authority test-blockchains
+        if poa is True:
+            w3 = self.interface.w3
```
Comment (michwill) [inline nucypher/utilities/sandbox/blockchain.py]: In recent web3, one can do `w3 = web3.auto.gethdev.w3` (although, manually injecting into `middleware_stack` should also continue to work... Until high velocity breaks it again!)

File: nucypher/utilities/sandbox/ursula.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    else:
+        starting_port = max(TEST_KNOWN_URSULAS_CACHE.keys()) + 1
+
+    federated_ursulas = set()
+    for port in range(starting_port, starting_port+quantity):
+
+        ursula = Ursula(is_me=True,
+                        db_name="test-{}".format(port),
+                        rest_host="localhost",
+                        rest_port=port + 100,
```
Comment (michwill) [inline nucypher/utilities/sandbox/ursula.py]: If we don't have a DHT anymore, why `+ 100`?

File: nucypher/config/constants.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/config/constants.py]: I'm using this hidden file as a 'template' to write new configuration file via the CLI (perhaps interactively at some point) and a method: `NodeConfiguration.write_default_config_file(filepath=/my/config/file/path/)`.  If I understand correctly, if we ship a required default config file then we'll need to install it with `setup.py`, but for now it's optional.

File: nucypher/config/node.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/config/node.py]: Yes - Fixed upstream.

File: nucypher/utilities/sandbox/blockchain.py (hunk 3 above)
Comment (KPrasch) [reply nucypher/utilities/sandbox/blockchain.py]: This is only for supporting test runs with geth... at the moment - Ill look at the `web3.auto.gethdev.w3` API.
```
        # For use with Proof-Of-Authority test-blockchains
        if poa is True:
            w3 = self.interface.w3
            w3.middleware_stack.inject(geth_poa_middleware, layer=0)
```

File: nucypher/utilities/sandbox/ursula.py (hunk 4 above)
Comment (KPrasch) [reply nucypher/utilities/sandbox/ursula.py]: Hmm, good point. I'll change it.

File: nucypher/config/node.py [hunk 5]
```diff
@@ -1,106 +1,100 @@
...
+            except FileExistsError:
+                raise
+            except FileNotFoundError:
+                raise
+
+        #
+        # Create Config Subdirectories
+        #
 
+        os.mkdir(self.keyring_dir, mode=0o700)               # keyring
```
Comment (tuxxy) [inline nucypher/config/node.py]: :+1:

File: nucypher/utilities/sandbox/middleware.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    def put_treasure_map_on_node(self, node, map_id, map_payload):
+        mock_client = self._get_mock_client_by_ursula(node)
+        response = mock_client.post("http://localhost/treasure_map/{}".format(map_id),
+                                    data=map_payload, verify=False)
+        return response
+
+
+class EvilMiddleWare(MockRestMiddleware):
+    """
+    Middleware for assholes.
```
Comment (tuxxy) [inline nucypher/utilities/sandbox/middleware.py]: :joy:

File: nucypher/utilities/sandbox/ursula.py (hunk 4 above)
Comment (KPrasch) [reply nucypher/utilities/sandbox/ursula.py]: Changed upstream (to avoid additional rebases).

**Pull request merged**